### PR TITLE
Add /poems route

### PR DESCRIPTION
**Pull request by HinataIzayoi**
Add /v1/poems/create to allow User to create poem.
Add /v1/poems/edit to allow User to edit poem.
Add /v1/poems/delete to allow User to delete poem.
Add /v1/poems/like to allow User to like poem.
Add /v1/poems/unlike to allow User to unlike poem.
Add /v1/poems/visit to allow User or visitor to visit poem.

#10 #11 #13  #17 #19 #50

### CODE REVIEW

File: controllers/poemController.js [hunk 1]
```diff
@@ -0,0 +1,174 @@
...
+    poem: params.poemId,
+  });
+
+  const poem = await Poem.findById(params.poemId);
+
+  // Update viewCount for Poem and User
+  await Poem.findByIdAndUpdate(params.poemId,
+      {$inc: {viewCount: 1}});
+
+  await User.findByIdAndUpdate(poem.author,
```
Comment (Kailang) [inline controllers/poemController.js]: Parallelize.

File: controllers/userController.js [hunk 2]
```diff
@@ -104,7 +105,7 @@ exports.unfollow = async function(params) {
   // if not following unfollowId
   const followRelationCount = await UserFollowUser.find({follower: userId, following: params.unfollowId}).count();
   if (followRelationCount === 0) {
-    return apiError(FORBIDDEN);
+    return apiError(BAD_REQUEST);
```
Comment (Kailang) [inline controllers/userController.js]: Revert.

File: models/Poem.js [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+const ObjectId = mongoose.Schema.Types.ObjectId;
+
+module.exports = mongoose.model('Poem', {
+  author: ObjectId,
+  title: String,
+  body: String,
+  align: String,
+  writtenDate: Date,
+  lastEditDate: Date,
+  privacy: String,
```
Comment (Kailang) [inline models/Poem.js]: visibility.
Change this to a String field with enum validator:
```
visibility: {type: String, enum: [PUBLIC, COMMUNITY, PRIVATE]}
```

File: models/Poem.js [hunk 4]
```diff
@@ -0,0 +1,14 @@
+const mongoose = require('mongoose');
+const ObjectId = mongoose.Schema.Types.ObjectId;
+
+module.exports = mongoose.model('Poem', {
+  author: ObjectId,
```
Comment (Kailang) [inline models/Poem.js]: authorId.

File: models/Poem.js [hunk 5]
```diff
@@ -0,0 +1,14 @@
...
+  title: String,
+  body: String,
+  align: String,
```
Comment (Kailang) [inline models/Poem.js]: Where are the constants for align?

File: models/UserLikePoem.js [hunk 6]
```diff
@@ -0,0 +1,7 @@
+const mongoose = require('mongoose');
+const ObjectId = mongoose.Schema.Types.ObjectId;
+
+module.exports = mongoose.model('UserLikePoem', {
+  user: ObjectId,
```
Comment (Kailang) [inline models/UserLikePoem.js]: userId.

File: models/UserLikePoem.js [hunk 7]
```diff
@@ -0,0 +1,7 @@
...
+  poem: ObjectId,
```
Comment (Kailang) [inline models/UserLikePoem.js]: poemId.

File: models/UserVisitPoem.js [hunk 8]
```diff
@@ -0,0 +1,7 @@
+const mongoose = require('mongoose');
+const ObjectId = mongoose.Schema.Types.ObjectId;
+
+module.exports = mongoose.model('UserVisitPoem', {
+  user: ObjectId,
+  poem: ObjectId,
```
Comment (Kailang) [inline models/UserVisitPoem.js]: userId.
poemId.

File: controllers/userController.js [hunk 9]
```diff
@@ -118,3 +119,15 @@ exports.unfollow = async function(params) {
 
   return apiSuccess();
 };
+
+exports.detail = async function(params) {
```
Comment (Kailang) [inline controllers/userController.js]: Move changes of userController to a new branch and submit another PR.

File: controllers/poemController.js [hunk 10]
```diff
@@ -0,0 +1,174 @@
...
+    body: params.body,
+    writtenDate: params.date,
+    lastEditDate: params.date,
+    privacy: params.privacy,
+    align: params.align,
+    likeCount: 0,
+    viewCount: 0,
+  });
+
+  await User.findByIdAndUpdate(userId, {
```
[the remark this answers is not in the document]
Comment (HinataIzayoi) [reply controllers/poemController.js]: Done.

File: controllers/poemController.js [hunk 11]
```diff
@@ -0,0 +1,174 @@
...
+};
+
+exports.edit = async function(params) {
+  const userId = getUserId(params.token);
+  const poem = await Poem.findById(params.poemId);
+  if (!poem) {
+    return apiError(NOT_FOUND);
+  }
+
+  if (userId != poem.author) {
```
[the remark this answers is not in the document]
Comment (HinataIzayoi) [reply controllers/poemController.js]: Done.

File: controllers/poemController.js [hunk 12]
```diff
@@ -0,0 +1,174 @@
...
+    $set: {
+      title: params.title,
+      body: params.body,
+      lastEditDate: params.date,
+      privacy: params.privacy,
+      align: params.align,
+    },
+  });
+
+  await User.findByIdAndUpdate(userId, {
```
[the remark this answers is not in the document]
Comment (HinataIzayoi) [reply controllers/poemController.js]: Done.

File: controllers/poemController.js [hunk 13]
```diff
@@ -0,0 +1,174 @@
...
+};
+
+exports.delete = async function(params) {
+  const userId = getUserId(params.token);
+  const poem = await Poem.findById(params.poemId);
+  if (!poem) {
+    return apiError(NOT_FOUND);
+  }
+
+  if (userId != poem.author) {
```
[the remark this answers is not in the document]
Comment (HinataIzayoi) [reply controllers/poemController.js]: Done.

File: controllers/poemController.js [hunk 14]
```diff
@@ -0,0 +1,174 @@
...
+  if (existingCount > 0) {
+    return apiError(BAD_REQUEST);
+  }
+
+  await UserLikePoem.create({
+    user: userId,
+    poem: params.poemId,
+  });
+
+  await Poem.findByIdAndUpdate(params.poemId,
```
[the remark this answers is not in the document]
Comment (HinataIzayoi) [reply controllers/poemController.js]: Done.

File: controllers/poemController.js [hunk 15]
```diff
@@ -0,0 +1,174 @@
...
+
+  // If did not like poem
+  const likeRelationCount = await UserLikePoem.find({user: userId, poem: params.poemId}).count();
+  if (likeRelationCount === 0) {
+    return apiError(BAD_REQUEST);
+  }
+  await UserLikePoem.deleteMany({user: userId, poem: params.poemId});
+
+  // decrement likeCount
+  await Poem.findByIdAndUpdate(params.poemId,
```
[the remark this answers is not in the document]
Comment (HinataIzayoi) [reply controllers/poemController.js]: Done.

File: controllers/poemController.js (hunk 1 above)
Comment (HinataIzayoi) [reply controllers/poemController.js]: Done.

File: controllers/userController.js (hunk 2 above)
Comment (HinataIzayoi) [reply controllers/userController.js]: Done.

File: models/Poem.js (hunk 3 above)
Comment (HinataIzayoi) [reply models/Poem.js]: Done.

File: models/Poem.js (hunk 4 above)
Comment (HinataIzayoi) [reply models/Poem.js]: Done.

File: models/UserLikePoem.js (hunk 6 above)
Comment (HinataIzayoi) [reply models/UserLikePoem.js]: Done.

File: models/UserLikePoem.js (hunk 7 above)
Comment (HinataIzayoi) [reply models/UserLikePoem.js]: Done.

File: models/UserVisitPoem.js (hunk 8 above)
Comment (HinataIzayoi) [reply models/UserVisitPoem.js]: Done.

File: controllers/userController.js (hunk 9 above)
Comment (HinataIzayoi) [reply controllers/userController.js]: Done.

File: models/Poem.js (hunk 5 above)
Comment (HinataIzayoi) [reply models/Poem.js]: Acknowledged. The constants are in utils.js.
